### PR TITLE
feat(apollo-client): introduce apollo client

### PR DESCRIPTION
**Pull request by tglatt**
- [x] invalider le cookie à la déconnexion
- [x] Isoler dans un fichier setToken, getToken
- [x] isoler la query gql pour les composants

### CODE REVIEW

**Comment by tglatt** [comment]
@kwiss, The PR is not ready but have a look if you want!

File: packages/app/src/components/service/service-mesure/MesureList.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import gql from "graphql-tag";
+import React from "react";
+import { Query } from "react-apollo";
+
+export const mesuresQuery = gql`
```
Comment (kwiss) [inline packages/app/src/components/service/service-mesure/MesureList.js]: Put that in a dedicated folder

```
/graphql
  /queries
  /mutations
```

File: packages/app/src/components/loginComponents/LoginForm.js [hunk 2]
```diff
@@ -164,6 +164,9 @@ class LoginForm extends React.Component {
     // Saves user token to localStorage
     // todo: move to more secure location
     localStorage.setItem("id_token", idToken);
+    document.cookie = cookie.serialize("token", idToken, {
```
Comment (kwiss) [inline packages/app/src/components/loginComponents/LoginForm.js]: we can only use localstorage if we don't do isomorphic

**Comment by kwiss** [comment]
i don't think we need isomorphic at all

File: packages/app/src/lib/with-apollo-client.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import PropTypes from "prop-types";
+import React from "react";
+import { getDataFromTree } from "react-apollo";
+import initApollo from "./init-apollo";
+
+function parseCookies(req, options = {}) {
+  return cookie.parse(req ? req.headers.cookie || "" : document.cookie, options);
+}
+
+export default App => {
```
Comment (kwiss) [inline packages/app/src/lib/with-apollo-client.js]: it's a confusing naming maybe something apollo related should be better ? it's not real react hox so naming with is wrong in my opinion

Comment (tglatt) [reply packages/app/src/lib/with-apollo-client.js]: `export default withApolloClient(MyApp);`

File: packages/api/model/User.js [hunk 4]
```diff
@@ -61,6 +61,26 @@ class User extends Model {
     return this.roles.map(el => el.name).concat("user");
   }
 
+  getDefaultRole() {
+    for (const role of this.roles) {
```
Comment (douglasduteil) [inline packages/api/model/User.js]: Equal to :
`(roles.find((role)=>tierList.includes(role.name)) || {}).name`

File: packages/app/src/business/AuthService.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { storageService } from "../util";
+
+export const authService = {
+  logout: () => {
```
Comment (douglasduteil) [inline packages/app/src/business/AuthService.js]: `logout() {`

File: packages/app/src/components/communComponents/Navigation.js [hunk 6]
```diff
@@ -1,12 +1,18 @@
...
-const doLogout = () => {
-  // Clear user token and profile data from localStorage
-  localStorage.removeItem("id_token");
-  ReactPiwik.push(["resetUserId"]);
-  Router.push("/login");
+const doLogout = apolloClient => {
+  authService.logout();
+  // Force a reload of all the current queries now that the user is
+  // logged in, so we don't accidentally leave any state around.
+  apolloClient.cache.reset().then(() => {
```
Comment (douglasduteil) [inline packages/app/src/components/communComponents/Navigation.js]: async/await ?

File: packages/app/src/lib/apollo/with-apollo-client.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        apolloState
+      };
+    }
+
+    constructor(props) {
+      super(props);
+      // `getDataFromTree` renders the component first, the client is passed off as a property.
+      // After that rendering is done using Next's normal rendering pipeline
+      this.apolloClient = initApollo(props.apolloState, {
+        getToken: () => {
```
Comment (douglasduteil) [inline packages/app/src/lib/apollo/with-apollo-client.js]:
```
      this.apolloClient = initApollo(props.apolloState, {
        getToken() {
          return parseCookies().token;
        }
      });

// or 

      this.apolloClient = initApollo(props.apolloState, {
        getToken: () => parseCookies().token
      });
```

File: packages/app/src/util/storage/StorageService.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  removeLogin: () => {
+    localStorage.removeItem(LOGIN_KEY);
+  },
+  setLogin: login => {
+    localStorage.setItem(LOGIN_KEY, login);
+  },
+  setToken: token => {
+    localStorage.setItem(ID_TOKEN_KEY, token);
+    document.cookie = cookie.serialize(TOKEN_KEY, token, {
+      maxAge: 30 * 24 * 60 * 60 // 30 days
```
Comment (douglasduteil) [inline packages/app/src/util/storage/StorageService.js]: Can be a constant : `30_DAYS_TIME` or `30_DAYS_DURATION`

File: packages/app/src/lib/apollo/with-apollo-client.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      const { req, res } = ctx;
+
+      const apollo = initApollo(
+        {},
+        {
+          getToken: () => parseCookies(req).token
+        }
+      );
+
+      context.ctx.apolloClient = apollo;
```
Comment (douglasduteil) [inline packages/app/src/lib/apollo/with-apollo-client.js]: ctx.apolloClient = apollo;

**Pull request merged**